### PR TITLE
Add initial traefik support

### LINKED ISSUE #93: Use traefik container to proxy all web services
It would be ideal to actually proxy all the container web traffic via a single web proxy to make it simpler to expose ports.

We could look at using this:  https://github.com/containous/traefik

### PR DESCRIPTION
**Pull request by leifmadsen**
Adding initial traefik support to start wrapping our web services
with a reverse proxy to keep the number of ports being exposed much
simpler.

Closes #93

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -2,6 +2,16 @@ version: '2'
 
 # docker-compose file to launch jenkins_master/ELK stack instead of VM.
 services:
+  proxy:
+    image: traefik
+    command: --web --docker --docker.domain=${PROXY_DOMAIN} --logLevel=${PROXY_LOG_LEVEL}
+    ports:
+      - "80:80"
+      - "8080:8080"
+      - "443:443"
```
Comment (dougbtv) [inline docker-compose.yml]: Looks like there's a conflicting 443 allocation for Kibana later in the file (I wouldn't've found it, but, Travis did!)

**Review by dougbtv — CHANGES_REQUESTED** [review]

**Comment by leifmadsen** [comment]
Currently a WIP, please do not merge.

**Comment by s1061123** [comment]
Hi Leif, 

I'm wondering whether traefik makes network easier or not in our docker-compose case.
Currently our docker-compose scenario is mainly for one container host hence nothing is different between with traefic and without traefic. Without traefic, docker-compose expose works fine. Traefic is very useful in case of multi-host container environment, such as docker-swarm, k8s and mesos, so your plan is good for future use case scenario.

So how about to make separate docker-compose.yaml for 

- one container host scenario  (without traefic)
- multi container-host scenario (with traefic)?

File: docker-compose.yml (hunk 1 above)
Comment (leifmadsen) [reply docker-compose.yml]: Fixed!

**Comment by leifmadsen** [comment]
@s1061123 (I know we talked today and think we're on the same page, but for historical reference, I'll comment below to your questions above.)

I think for now traefik doesn't really add any complexity that makes it harder to deploy TOAD than it is now, in fact I think it makes it bit simpler since you don't need to know the specific ports for connecting to the services. Right now, you would connect to Jenkins via:  http://toad.tld/jenkins and everything would just route.

If you don't want to use a hostname (specified via your `/etc/hosts` file for example), then you could override the `.env` file with your own `PROXY_DOMAIN` that points at the IP address that you want to listen on (untested).

Overall, I think while it doesn't add a lot to the system as it stands, it is a good tool for helping to future proof a dynamic scaling scenario, and also for the cases where we want to deal with life cycle management, allowing for two containers to live simultaneously behind the proxy, in order to upgrade the software running in that container.

**Review by dougbtv — APPROVED** [review]

**Pull request merged**